File: src/nuc/Node.js
const { graph } = require("../graph");
const transaction = require("../transaction");

let sequence = 0;

class Node {
  constructor(key) {
    this.key = key;
    this.next = {};
    this.previous = {};
    this.sequence = sequence++;
  }

  before() {}
  run() {}
  beforeGraph() {}
  graph() {}

  static register(key, node) {
    transaction.register(graph, key, node);
  }

  static replace(sourceKey, targetNode) {
    transaction.register(targetNode.block, graph[sourceKey].block);

    for (let node in graph[sourceKey].next) {
      transaction.register(targetNode.next, node, graph[sourceKey].next[node]);
      transaction.register(graph[sourceKey].next, node, undefined);
    }

    for (let node in graph[sourceKey].previous) {
      transaction.register(graph[node].next, sourceKey, undefined);
    }

    transaction.register(graph, sourceKey, targetNode);
  }

  static direct(sourceKey, targetKey, targetNode) {
    transaction.register(graph[sourceKey].next, targetKey, targetNode);
    transaction.register(targetNode.previous, sourceKey, graph[targetKey]);
  }
}

module.exports = Node;

File: cpp/src/nuc/NODE.cpp
#include "nucleoid/nuc/NODE.hpp"
#include "nucleoid/Graph.hpp"
#include "nucleoid/Transaction.hpp"
#include <stdexcept>

namespace nucleoid {

std::atomic<int> NODE::s_sequence{0};

NODE::NODE(const std::string& k)
    : key(k), sequence(s_sequence.fetch_add(1)) {}

void NODE::registerNode(const std::string& key, NODE* node) {
    Graph::instance().store(key, node);
}

void NODE::replaceNode(const std::string& srcKey, NODE* target) {
    NODE* src = Graph::instance().retrieve(srcKey);
    if (!src) return;

    // Transfer all outgoing edges from src to target.
    for (auto& [k, v] : src->next) {
        target->next[k] = v;
    }
    src->next.clear();

    // Remove incoming edges that pointed to src.
    for (auto& [k, v] : src->previous) {
        NODE* pred = Graph::instance().retrieve(k);
        if (pred) pred->next.erase(srcKey);
    }

    Graph::instance().store(srcKey, target);
}

void NODE::directNode(const std::string& srcKey,
                      const std::string& tgtKey, NODE* target) {
    NODE* src = Graph::instance().retrieve(srcKey);
    if (!src) return;
    src->next[tgtKey]      = target;
    target->previous[srcKey] = src;
}

} // namespace nucleoid
